File: services/api/src/main/java/com/launchly/workspace/controllers/MemberController.java
package com.launchly.workspace.controllers;

import com.launchly.auth.entities.User;
import com.launchly.auth.repositories.UserRepository;
import com.launchly.common.security.AuthContext;
import com.launchly.workspace.entities.WorkspaceMember;
import com.launchly.workspace.enums.Role;
import com.launchly.workspace.repositories.WorkspaceMemberRepository;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

@RestController
@RequestMapping("/api/members")
public class MemberController {
    private final WorkspaceMemberRepository memberRepository;
    private final UserRepository userRepository;

    public MemberController(WorkspaceMemberRepository memberRepository, UserRepository userRepository) {
        this.memberRepository = memberRepository;
        this.userRepository = userRepository;
    }

    @GetMapping
    public ResponseEntity<List<MemberResponse>> list() {
        String workspaceId = AuthContext.workspaceId();
        List<WorkspaceMember> members = memberRepository.findByWorkspaceId(workspaceId);

        List<String> userIds = members.stream().map(WorkspaceMember::getUserId).collect(Collectors.toList());
        Map<String, User> userMap = userRepository.findAllById(userIds).stream()
                .collect(Collectors.toMap(User::getId, u -> u));

        List<MemberResponse> result = members.stream().map(m -> {
            User user = userMap.get(m.getUserId());
            String account = user != null ? user.getAccount() : "unknown";
            String displayName = user != null ? user.getDisplayName() : null;
            return new MemberResponse(m.getId(), m.getUserId(), account, displayName, m.getRole().name(), m.getCreatedAt().toString());
        }).collect(Collectors.toList());

        return ResponseEntity.ok(result);
    }

    @PutMapping("/{id}/role")
    @PreAuthorize("hasRole('OWNER')")
    public ResponseEntity<Void> updateRole(@PathVariable String id, @RequestBody Map<String, String> body) {
        String workspaceId = AuthContext.workspaceId();
        WorkspaceMember member = memberRepository.findById(id).orElse(null);
        if (member == null || !member.getWorkspaceId().equals(workspaceId)) {
            return ResponseEntity.notFound().build();
        }

        String newRole = body.get("role");
        if (newRole == null) {
            return ResponseEntity.badRequest().build();
        }

        try {
            member.setRole(Role.valueOf(newRole));
            memberRepository.save(member);
            return ResponseEntity.ok().build();
        } catch (IllegalArgumentException e) {
            return ResponseEntity.badRequest().build();
        }
    }

    @DeleteMapping("/{id}")
    @PreAuthorize("hasRole('OWNER')")
    public ResponseEntity<Void> remove(@PathVariable String id) {
        String workspaceId = AuthContext.workspaceId();
        WorkspaceMember member = memberRepository.findById(id).orElse(null);
        if (member == null || !member.getWorkspaceId().equals(workspaceId)) {
            return ResponseEntity.notFound().build();
        }

        // Prevent removing the last owner
        if (member.getRole() == Role.OWNER) {
            long ownerCount = memberRepository.findByWorkspaceId(workspaceId).stream()
                    .filter(m -> m.getRole() == Role.OWNER)
                    .count();
            if (ownerCount <= 1) {
                return ResponseEntity.status(409).build();
            }
        }

        memberRepository.delete(member);
        return ResponseEntity.ok().build();
    }

    public record MemberResponse(String id, String userId, String account, String displayName, String role, String createdAt) {}
}

File: services/api/src/workspace/member.controller.ts
import { Controller, Get, Put, Delete, Param, Body, ForbiddenException, ConflictException } from '@nestjs/common';
import { PrismaService } from '../common/prisma/prisma.service';
import { CurrentUser, AuthPrincipal } from '../common/decorators/current-user.decorator';
import { Roles } from '../common/decorators/roles.decorator';

@Controller('members')
export class MemberController {
  constructor(private readonly prisma: PrismaService) {}

  @Get()
  async list(@CurrentUser() user: AuthPrincipal) {
    const members = await this.prisma.workspaceMember.findMany({
      where: { workspaceId: user.workspaceId },
      include: { user: true },
    });
    return members.map(m => ({
      id: m.id,
      userId: m.userId,
      account: m.user.account,
      displayName: m.user.displayName,
      role: m.role,
      createdAt: m.createdAt.toISOString(),
    }));
  }

  @Roles('OWNER')
  @Put(':id/role')
  async updateRole(@Param('id') id: string, @Body('role') newRole: string, @CurrentUser() user: AuthPrincipal) {
    const member = await this.prisma.workspaceMember.findFirst({
      where: { id, workspaceId: user.workspaceId },
    });
    if (!member) throw new ForbiddenException('成员不存在');

    const validRoles = ['OWNER', 'ADMIN', 'DEVELOPER', 'TESTER', 'VIEWER'];
    if (!validRoles.includes(newRole)) throw new ForbiddenException('无效角色');

    await this.prisma.workspaceMember.update({
      where: { id },
      data: { role: newRole },
    });
    return { success: true };
  }

  @Roles('OWNER')
  @Delete(':id')
  async remove(@Param('id') id: string, @CurrentUser() user: AuthPrincipal) {
    const member = await this.prisma.workspaceMember.findFirst({
      where: { id, workspaceId: user.workspaceId },
    });
    if (!member) throw new ForbiddenException('成员不存在');

    // Prevent removing the last owner
    if (member.role === 'OWNER') {
      const ownerCount = await this.prisma.workspaceMember.count({
        where: { workspaceId: user.workspaceId, role: 'OWNER' },
      });
      if (ownerCount <= 1) throw new ConflictException('不能移除最后一个 Owner');
    }

    await this.prisma.workspaceMember.delete({ where: { id } });
    return { success: true };
  }
}
